File: scraper/github_scraper.py
"""
GitHub Scraper — fetches live data from the GitHub API.

Every number returned has a `_scraped_at` timestamp proving when it was
verified.  Nothing is cached across runs; if a request fails the field
is omitted rather than backfilled with stale data.
"""

import json
import os
import urllib.request
import urllib.error
from datetime import datetime, timezone
from typing import Optional


# The 5 repos this operator instance tracks.
TARGET_REPOS: list[dict[str, str]] = [
    {"owner": "BlackRoad-OS", "repo": "blackroad"},
    {"owner": "BlackRoad-OS", "repo": "blackroad-os"},
    {"owner": "BlackRoad-OS", "repo": ".github"},
    {"owner": "BlackRoad-OS", "repo": "blackroad-os-web"},
    {"owner": "BlackRoad-OS", "repo": "operator"},
]

API_BASE = "https://api.github.com"


def _headers() -> dict[str, str]:
    """Build request headers, including auth token if available."""
    h = {
        "Accept": "application/vnd.github+json",
        "User-Agent": "BlackRoad-OS-Operator-Scraper/1.0",
    }
    token = os.environ.get("GITHUB_TOKEN") or os.environ.get("GH_TOKEN")
    if token:
        h["Authorization"] = f"Bearer {token}"
    return h


def _get_json(url: str) -> Optional[dict]:
    """GET a URL and return parsed JSON, or None on failure."""
    req = urllib.request.Request(url, headers=_headers())
    try:
        with urllib.request.urlopen(req, timeout=15) as resp:
            return json.loads(resp.read().decode())
    except (urllib.error.URLError, urllib.error.HTTPError, OSError, json.JSONDecodeError):
        return None


def _link_header_last_page(url: str) -> Optional[int]:
    """Fetch HEAD/GET with per_page=1 and parse Link header for last page count."""
    req = urllib.request.Request(f"{url}?per_page=1&state=all", headers=_headers())
    try:
        with urllib.request.urlopen(req, timeout=15) as resp:
            link = resp.headers.get("Link", "")
            for part in link.split(","):
                if 'rel="last"' in part:
                    # <https://...?per_page=1&page=42>; rel="last"
                    page_str = part.split("page=")[-1].split(">")[0]
                    return int(page_str)
    except (urllib.error.URLError, urllib.error.HTTPError, OSError, ValueError):
        pass
    return None


def scrape_repo(owner: str, repo: str) -> dict:
    """
    Scrape a single repo.  Returns only fields we actually got back
    from the API right now — never fills in from old data.
    """
    now = datetime.now(timezone.utc).isoformat()
    result: dict = {
        "owner": owner,
        "repo": repo,
        "_scraped_at": now,
    }

    # ── core repo metadata ──
    data = _get_json(f"{API_BASE}/repos/{owner}/{repo}")
    if data and isinstance(data, dict):
        result["stars"] = data.get("stargazers_count")
        result["forks"] = data.get("forks_count")
        result["open_issues_count"] = data.get("open_issues_count")
        result["language"] = data.get("language")
        result["default_branch"] = data.get("default_branch")
        result["description"] = data.get("description")
        result["homepage"] = data.get("homepage")
        result["topics"] = data.get("topics", [])
        result["license"] = (data.get("license") or {}).get("spdx_id")
        result["created_at"] = data.get("created_at")
        result["updated_at"] = data.get("updated_at")
        result["pushed_at"] = data.get("pushed_at")
        result["size_kb"] = data.get("size")
        result["archived"] = data.get("archived")
        result["has_wiki"] = data.get("has_wiki")
        result["has_pages"] = data.get("has_pages")
        result["visibility"] = data.get("visibility")
        result["_api_ok"] = True
    else:
        result["_api_ok"] = False

    # ── contributors count ──
    contribs = _get_json(f"{API_BASE}/repos/{owner}/{repo}/contributors?per_page=1&anon=true")
    if isinstance(contribs, list):
        # Use link header for real count
        count = _link_header_last_page(f"{API_BASE}/repos/{owner}/{repo}/contributors")
        result["contributors_count"] = count if count else len(contribs)

    # ── open PRs count (separate from issues) ──
    prs = _get_json(f"{API_BASE}/repos/{owner}/{repo}/pulls?state=open&per_page=1")
    if isinstance(prs, list):
        pr_count = _link_header_last_page(f"{API_BASE}/repos/{owner}/{repo}/pulls")
        result["open_prs"] = pr_count if pr_count else len(prs)

    # ── branches count ──
    branches = _get_json(f"{API_BASE}/repos/{owner}/{repo}/branches?per_page=1")
    if isinstance(branches, list):
        br_count = _link_header_last_page(f"{API_BASE}/repos/{owner}/{repo}/branches")
        result["branches_count"] = br_count if br_count else len(branches)

    # ── latest release ──
    release = _get_json(f"{API_BASE}/repos/{owner}/{repo}/releases/latest")
    if release and isinstance(release, dict) and "tag_name" in release:
        result["latest_release"] = release["tag_name"]
        result["latest_release_date"] = release.get("published_at")

    # Strip None values — if we didn't get it, don't report it
    return {k: v for k, v in result.items() if v is not None}


def scrape_all() -> list[dict]:
    """Scrape all target repos. Returns list of verified results."""
    return [scrape_repo(r["owner"], r["repo"]) for r in TARGET_REPOS]


def scrape_to_json(output_path: str = "scraped_data.json") -> str:
    """Scrape all repos and write results to JSON file. Returns the path."""
    results = scrape_all()
    with open(output_path, "w") as f:
        json.dump(results, f, indent=2)
    return output_path


if __name__ == "__main__":
    print(json.dumps(scrape_all(), indent=2))

File: src/scraper/github-scraper.js
'use strict';

const axios = require('axios');

/**
 * GitHubScraper - Scrapes and validates real metrics from GitHub repos.
 * Every number returned is verified at scrape time. Nothing cached, nothing assumed.
 */
class GitHubScraper {
  constructor(options = {}) {
    this.token = options.token || process.env.GITHUB_TOKEN || null;
    this.baseUrl = options.baseUrl || 'https://api.github.com';
    this.timeout = options.timeout || 15000;
    this.retries = options.retries || 3;
    this.retryDelay = options.retryDelay || 2000;
  }

  _headers() {
    const h = {
      'Accept': 'application/vnd.github.v3+json',
      'User-Agent': 'BlackRoad-Operator/1.0',
    };
    if (this.token) h['Authorization'] = `Bearer ${this.token}`;
    return h;
  }

  async _request(url) {
    let lastError;
    for (let attempt = 0; attempt < this.retries; attempt++) {
      try {
        const resp = await axios.get(url, {
          headers: this._headers(),
          timeout: this.timeout,
        });
        return { data: resp.data, status: resp.status, headers: resp.headers };
      } catch (err) {
        lastError = err;
        if (attempt < this.retries - 1) {
          const delay = this.retryDelay * Math.pow(2, attempt);
          await new Promise(r => setTimeout(r, delay));
        }
      }
    }
    return { error: lastError.message, status: lastError.response?.status || 0 };
  }

  async scrapeRepo(owner, repo) {
    const ts = new Date().toISOString();
    const url = `${this.baseUrl}/repos/${owner}/${repo}`;
    const result = await this._request(url);

    if (result.error) {
      return {
        owner, repo, scraped_at: ts, status: 'error',
        error: result.error, http_status: result.status,
        verified: false,
      };
    }

    const d = result.data;
    return {
      owner, repo, scraped_at: ts, status: 'ok',
      verified: true,
      metrics: {
        stars: d.stargazers_count,
        forks: d.forks_count,
        watchers: d.subscribers_count,
        open_issues: d.open_issues_count,
        size_kb: d.size,
        default_branch: d.default_branch,
        language: d.language,
        created_at: d.created_at,
        updated_at: d.updated_at,
        pushed_at: d.pushed_at,
        archived: d.archived,
        disabled: d.disabled,
        visibility: d.visibility,
        has_issues: d.has_issues,
        has_wiki: d.has_wiki,
        has_pages: d.has_pages,
        license: d.license?.spdx_id || null,
        description: d.description,
        topics: d.topics || [],
      },
    };
  }

  async scrapeRepoCommits(owner, repo, opts = {}) {
    const ts = new Date().toISOString();
    const perPage = opts.perPage || 30;
    const sha = opts.branch || '';
    let url = `${this.baseUrl}/repos/${owner}/${repo}/commits?per_page=${perPage}`;
    if (sha) url += `&sha=${sha}`;

    const result = await this._request(url);
    if (result.error) {
      return { owner, repo, scraped_at: ts, status: 'error', error: result.error, verified: false };
    }

    const commits = result.data.map(c => ({
      sha: c.sha,
      message: c.commit.message.split('\n')[0],
      author: c.commit.author.name,
      date: c.commit.author.date,
    }));

    return {
      owner, repo, scraped_at: ts, status: 'ok', verified: true,
      commit_count: commits.length,
      latest_commit: commits[0] || null,
      commits,
    };
  }

  async scrapeRepoWorkflows(owner, repo) {
    const ts = new Date().toISOString();
    const url = `${this.baseUrl}/repos/${owner}/${repo}/actions/workflows`;
    const result = await this._request(url);

    if (result.error) {
      return { owner, repo, scraped_at: ts, status: 'error', error: result.error, verified: false };
    }

    const workflows = (result.data.workflows || []).map(w => ({
      id: w.id,
      name: w.name,
      state: w.state,
      path: w.path,
      created_at: w.created_at,
      updated_at: w.updated_at,
    }));

    return {
      owner, repo, scraped_at: ts, status: 'ok', verified: true,
      total_workflows: result.data.total_count,
      workflows,
    };
  }

  async scrapeRepoLatestRuns(owner, repo, opts = {}) {
    const ts = new Date().toISOString();
    const perPage = opts.perPage || 10;
    const url = `${this.baseUrl}/repos/${owner}/${repo}/actions/runs?per_page=${perPage}`;
    const result = await this._request(url);

    if (result.error) {
      return { owner, repo, scraped_at: ts, status: 'error', error: result.error, verified: false };
    }

    const runs = (result.data.workflow_runs || []).map(r => ({
      id: r.id,
      name: r.name,
      status: r.status,
      conclusion: r.conclusion,
      branch: r.head_branch,
      event: r.event,
      created_at: r.created_at,
      updated_at: r.updated_at,
      run_started_at: r.run_started_at,
      html_url: r.html_url,
    }));

    return {
      owner, repo, scraped_at: ts, status: 'ok', verified: true,
      total_runs: result.data.total_count,
      runs,
    };
  }

  async scrapeRepoPulls(owner, repo, opts = {}) {
    const ts = new Date().toISOString();
    const state = opts.state || 'open';
    const perPage = opts.perPage || 30;
    const url = `${this.baseUrl}/repos/${owner}/${repo}/pulls?state=${state}&per_page=${perPage}`;
    const result = await this._request(url);

    if (result.error) {
      return { owner, repo, scraped_at: ts, status: 'error', error: result.error, verified: false };
    }

    const pulls = result.data.map(p => ({
      number: p.number,
      title: p.title,
      state: p.state,
      user: p.user.login,
      created_at: p.created_at,
      updated_at: p.updated_at,
      merged_at: p.merged_at,
      draft: p.draft,
    }));

    return {
      owner, repo, scraped_at: ts, status: 'ok', verified: true,
      count: pulls.length,
      pulls,
    };
  }

  async fullScrape(owner, repo) {
    const [repoData, commits, workflows, runs, pulls] = await Promise.all([
      this.scrapeRepo(owner, repo),
      this.scrapeRepoCommits(owner, repo),
      this.scrapeRepoWorkflows(owner, repo),
      this.scrapeRepoLatestRuns(owner, repo),
      this.scrapeRepoPulls(owner, repo),
    ]);

    return {
      owner, repo,
      scraped_at: new Date().toISOString(),
      verified: repoData.verified,
      repo: repoData,
      commits,
      workflows,
      runs,
      pulls,
    };
  }
}

module.exports = { GitHubScraper };
